### PR TITLE
update and reorganize docs for machine learning tutorial

### LINKED ISSUE #523: tools to develop mask for naive bayes classifier
Curious if you have a favorite tool for manually drawing a mask that can be used to the train the naive bayes classifier? Presumably the point of using the naive bayes classifier is that binary thresholding doesn't work well.

I can try the multiclass classifier with a csv of r,g,b values created from ImageJ too. Is there a way to record the pixel values of the pixels you click on with Pixel Inspector?

I think it would be a nice addition to the documentation if we had canonical methods for setting up these classifiers. happy to work on it but some guidance for existing practices would be helpful!

### PR DESCRIPTION
**Pull request by dschneiderch**
**Describe your changes**
I used the bayesian classifiers for the first time and added some details to the documentation about creating the rgb text files and reorganized. In particular I removed the last section because it felt redundant to the 2nd to last section and ultimately made it more confusing.

**Type of update**
* Update to documentation

**Associated issues**
addresses and closes #523

### CODE REVIEW

File: docs/machine_learning_tutorial.md [hunk 1]
```diff
@@ -14,12 +14,12 @@ The naive Bayes approach used here can be trained to label pixels as plant or ba
...
 
-To train the classifier, we need to label a relatively small set of images using a binary mask (just like above).
+To train the classifier, we need to label a relatively small set of images using a binary mask.
 We can use PlantCV to create a binary mask for a set of input images using the methods described in the 
-[VIS tutorial](vis_tutorial.md).
+[VIS tutorial](vis_tutorial.md). Alternatively, you can outline and create masks by hand.
 
 For the purpose of this tutorial, we assume we are in a folder containing two subfolders, one containing original RGB
-images, and one containing black and white masks that match the set of RGB images.
+images, and one containing black and white masks match the set of RGB images.
```
Comment (nfahlgren) [inline docs/machine_learning_tutorial.md:22]:
```suggestion
images, and one containing black and white masks that match the set of RGB images.
```

File: docs/machine_learning_tutorial.md [hunk 2]
```diff
@@ -133,9 +134,9 @@ def options():
     parser.add_argument("-o", "--outdir", help="Output directory for image files.", required=False)
     parser.add_argument("-r", "--result", help="result file.", required=False)
     parser.add_argument("-r2", "--coresult", help="result file.", required=False)
```
Comment (nfahlgren) [inline docs/machine_learning_tutorial.md:136]:
```suggestion
```

**Comment by dschneiderch** [comment]
I ended up writing an R script to convert my copy+pasted rgb values into columns for the train.py script. making those columns by hand was quite tedious. I would be happy to share but I guess it would be better as a python script if it were to be added to plantcv?

**Comment by nfahlgren** [comment]
What about adding the R script to the utils directory?

**Comment by nfahlgren** [comment]
I can actually remove most if not all of the scripts currently in utils.

**Comment by dschneiderch** [comment]
> I can actually remove most if not all of the scripts currently in utils.

confused. don't we need json2csv converter and sample_images ? those are the only scripts I see in utils/

**Comment by dschneiderch** [comment]
> 
> 
> > I can actually remove most if not all of the scripts currently in utils.
> 
> confused. don't we need json2csv converter and sample_images ? those are the only scripts I see in utils/

nevermind. thats plantcv/utils.
i've never looked at /utils

**Comment by nfahlgren** [comment]
Yeah, we used to have (or still do) a bunch of accessory utils. We converted the random sampler into a module for `plantcv-utils.py` but many of these others are not longer needed because they worked with the old SQLite outputs. But we can have other useful accessory scripts/tools that could really be in any language. These are not covered under the unit testing framework.

**Comment by dschneiderch** [comment]
script. this is script is run after you replace \t with \n in a text editor but I assume I could also do that part in the script.

```
#! Rscript
# Command-line script to convert seqential class rgb data to columns of classes
# Input:
#   fn = relative path to file with class rgb values in a single column
#
# File contents should look like this. No numbers in class names:
# Plant
# 0,255,0
# Background
# 255,255,255
# Necrosis
# 255,0,0
#
# Example: Rscript --vanilla scripts/prepare_bayes_classes.R "data/naive_bayes_training/bayes_rgb.tsv"

# Get command line arguments
args = commandArgs(trailingOnly = T)

# library(tidyverse)
# library(here)

fn = args[1]
# fn = here('data','naive_bayes_training','bayes_rgb.tsv')

#read each line of file
raw = readLines(fn)
# find lines with letters
hdr = grep("[[:alpha:]]",raw)
# add an additional entry for the last line in the file so we can keep our logic onsistent below
hdr_end = c(hdr,length(raw)+1)

# read chunks of file per class into named list
rgb=list()
for(i in seq_along(hdr)){
  j=i+1
  rgb[[i]] = raw[(hdr_end[i]+1):(hdr_end[j]-1)]
}
names(rgb) <- raw[hdr]

# find the max number of entries in a class and append NA to each list so length is the same
maxrows = max(sapply(rgb,length))
for(i in seq_along(rgb)){
  addnrows = maxrows-length(rgb[[i]])
  rgb[[i]] = c(rgb[[i]],rep('',addnrows))
}

# save dataframe of list as tab delimited file with columns
write.table(as.data.frame(rgb), file.path(dirname(fn),'bayes_rgb_trainingclasses.tsv'), row.names=F, sep='\t', quote = F)

```

**Comment by dschneiderch** [comment]
@nfahlgren  how does one find these scripts in an install? updating the doc again to include the script but do I just point them to the github link for the script?

**Comment by nfahlgren** [comment]
Oh right, good point. They are actually not (currently) included in a standard install, only if you had the source code.

Installed scripts get put into a global or user bin directory, so they need to be executable. They need to be added to an inventory in a couple of places.

In `setup.py` they get added to this list: `scripts=["plantcv-train.py", "plantcv-utils.py", "plantcv-workflow.py"]`.

They also need to be appended onto the `MANIFEST.in` file.

Alternatively we could write a Python module in the utils subpackage, which do you think is better?

**Comment by dschneiderch** [comment]
well i just updated the doc to include use of the R script for converting.

But it would probably be easier overall to keep to python:

I have `run_bayes_classifier.sh`
```
Rscript scripts/prepare_bayes_classes.R data/naive_bayes_training/bayes_rgb.tsv

plantcv-train.py naive_bayes_multiclass \
-f data/naive_bayes_training/bayes_rgb_trainingclasses.tsv \
-o data/naive_bayes_training/naive_bayes_pdfs.tsv
```

A plantcv-utils.py module would make sense to me:
```
python $CONDA_PREFIX/bin/plantcv-utils.py bayes_class_columns -i data/naive_bayes_training/bayes_rgb.tsv

plantcv-train.py naive_bayes_multiclass \
-f data/naive_bayes_training/bayes_rgb_trainingclasses.tsv \
-o data/naive_bayes_training/naive_bayes_pdfs.tsv
```

**Comment by dschneiderch** [comment]
oh! not needed
```
python $CONDA_PREFIX/bin/
```
PATH magic
😌

**Comment by dschneiderch** [comment]
I should also say I  need to work on somethign else atm but would keep the conversion to python on my radar unless someone else gets to it first

**Comment by nfahlgren** [comment]
Working on the Python version.

**Comment by nfahlgren** [comment]
@dschneiderch I didn't add to the documentation but pushed a new function and tests, see what you think. I'm not married to the names or anything, so feel free to suggest alternatives.

For the input file format I'm expecting two things. First, the class names are on their own line and start with the "#" character. Second, the user copies and pastes the tab-delimited RGB values from ImageJ below each class. It doesn't care how many columns/rows the RGB values sections have. And the 2+ classes do not need to have an equal number of RGB value triplets. Example from the test data:

```
#plant
96,154,72	95,153,72	91,155,71	91,160,70	90,155,67	92,152,66	92,157,70
98,156,71	97,155,72	95,153,70	93,156,66	93,153,66	92,155,65	94,158,69
101,159,70	99,161,70	98,156,68	96,155,67	94,155,67	93,155,68	94,156,69
98,159,69	97,157,67	97,156,67	96,156,67	97,153,68	96,156,68	97,153,70
97,156,71	97,150,69	97,153,71	98,158,71	98,158,70	98,160,70	100,156,72
94,157,72	96,154,72	99,149,73	99,156,75	100,160,72	100,157,70	102,151,73
89,156,73	94,160,72	99,154,72	100,154,72	101,157,71	101,155,70	101,156,71
52,99,27	52,106,32	53,105,35	55,107,34	53,103,33	50,96,31	50,89,30
52,102,32	55,105,32	54,106,32	54,106,32	51,101,31	49,93,31	48,91,30
54,99,30	59,103,31	57,104,33	54,104,33	52,104,32	47,96,31	46,91,30
53,99,29	57,103,32	55,104,33	55,103,33	50,101,33	47,95,31	47,89,30
53,97,29	55,102,31	55,104,33	54,104,31	50,103,32	47,95,30	48,87,30
53,99,29	55,102,30	52,105,32	52,105,32	50,100,30	49,94,30	48,90,29
53,98,29	53,101,30	52,104,30	51,103,31	51,96,33	50,91,31	48,87,30
48,85,32	48,89,33	49,90,35	50,91,34	51,90,33	54,93,36	54,93,37
45,84,32	45,86,32	46,87,32	48,89,34	52,86,38	55,91,40	55,95,41
48,85,32	47,85,31	49,87,28	52,90,36	54,89,42	55,91,43	55,92,43
48,84,31	48,85,32	52,83,34	55,89,39	55,91,42	55,91,43	56,91,45
50,89,30	51,87,34	52,87,38	55,89,41	55,91,42	56,92,44	60,92,44
52,94,36	51,90,37	52,87,39	52,89,42	55,90,44	59,90,46	64,91,46
53,89,41	53,90,41	53,89,41	53,92,44	56,90,47	59,88,47	65,95,47
54,104,39	56,104,38	59,106,41	57,105,43	54,104,40	54,103,35	56,101,39	58,99,41	59,99,41
52,102,40	54,102,38	58,104,40	54,105,43	51,104,40	54,102,38	58,102,39	59,99,40	59,102,40
55,102,41	56,102,38	60,103,39	56,104,40	53,104,40	54,103,39	54,104,39	56,102,37	58,106,40
56,99,41	59,100,38	63,102,39	58,103,39	54,103,39	53,100,40	51,102,38	54,101,38	58,104,40
54,100,40	56,101,38	60,103,37	58,103,37	54,104,39	54,103,40	54,101,39	56,102,36	59,105,40
53,102,38	56,99,37	58,103,39	57,101,40	56,103,40	57,106,41	57,102,40	58,99,39	61,104,41
53,104,38	54,104,38	56,108,40	57,103,41	59,100,41	58,102,41	56,99,41	59,102,41	61,104,41
53,103,38	54,102,40	54,103,40	57,101,41	61,102,40	57,102,40	55,102,41	57,106,41	61,107,41
55,99,40	56,99,41	55,101,41	57,103,40	59,106,40	58,103,39	57,102,40	59,106,41	63,111,41

#background
114,127,121	117,135,125	120,137,131	132,145,138	142,154,148	151,166,158	160,182,172
115,125,121	118,131,123	122,132,135	133,142,144	141,151,152	150,166,158	159,179,172
115,123,121	118,129,126	121,135,133	131,141,142	138,154,149	147,169,156	158,181,172
110,121,119	116,131,128	121,135,133	128,145,138	137,153,146	146,165,152	159,177,172
103,127,120	113,139,128	120,141,132	127,151,137	134,156,144	145,162,151	158,178,170
106,129,120	115,133,127	121,134,131	127,149,135	134,156,143	144,163,149	157,176,168
109,124,118	116,131,131	122,138,133	128,150,137	134,156,145	144,159,153	153,181,169
215,209,214	188,189,175	170,186,183	174,193,190	179,193,193	182,198,194	183,193,191
211,212,211	188,176,170	174,184,180	174,196,191	175,197,193	180,204,196	184,195,192
209,212,207	185,190,165	172,181,177	174,192,191	177,197,195	178,199,197	181,194,191
210,212,209	186,180,168	168,184,177	172,197,187	175,198,191	177,200,194	178,193,189
216,207,210	184,188,171	163,178,177	170,191,184	180,194,187	178,197,190	177,194,185
217,211,210	183,179,168	157,183,177	170,193,186	183,197,187	181,198,186	177,191,184
215,208,208	182,190,166	160,185,177	170,194,189	179,197,186	176,199,183	176,193,183
134,153,153	136,154,157	140,151,159	138,155,162	136,153,159	138,157,157	142,155,158
130,154,154	136,154,157	141,154,158	137,157,159	134,154,158	139,155,158	146,157,158
132,152,157	137,154,156	141,152,156	139,153,156	137,152,157	139,155,158	143,155,158
133,155,157	138,158,156	143,156,157	142,153,158	141,153,159	141,158,161	139,158,159
134,156,157	135,157,155	138,158,157	137,154,159	137,150,162	138,155,163	138,155,161
134,155,154	134,154,155	134,155,157	134,152,158	132,155,158	134,158,158	136,155,156
135,154,152	136,154,157	137,157,157	135,156,156	135,159,154	135,157,152	135,153,153
```

We don't have to use the "#" character, but some type of marker for the class names makes this function more straightforward.

**Comment by dschneiderch** [comment]
it looks so pythonic! i think i will forever be an R user that scratches his way through python code unpythonically. anyway, it worked well with my files.
It worked well except one minor thing in plantcv-utils (and plantcv-train) where I replaced `sys.exit(1)` with `raise SystemExit()`. sometimes i use `ipython` instead of `python` to run the scripts and `ipython` does not like `sys.exit(1)` and throws ugly error messages.

**Comment by dschneiderch** [comment]
> 
> 
> oh! not needed
> 
> ```
> python $CONDA_PREFIX/bin/
> ```
> 
> PATH magic
> 😌

actually, needed on Windows.

**Comment by dschneiderch** [comment]
Awesome thanks for your help.  I think this will make this tool easier for people to get into
________________________________

**Comment by nfahlgren** [comment]
Thanks for helping improve these tools for everyone @dschneiderch!

**Pull request merged**